### DISCUSSION TITLE
[rlvk] Vulkan support

### DISCUSSION THREAD

**Discussion by mandar1jn** [raysan5/raylib#2048] [Ideas]
I'm planning on adding Vulkan support to Raylib, but there are some things I need to know first.

The goal is to be able to compile for both OpenGL and Vulkan without having to change your program. 

How would you guys recommend I go about this?
Anything you'd like to see?

**Comment by mandar1jn** [comment]
So I need your opinion on something. Vulkan requires a engine version. Should that be the raylib version or the rlvk version? And there is also an application version parameter. Should that be something that just is always 1.0.0? Or should it be something that should be added to both rlgl and rlvk? And let rlgl just log it to the console? So developers can keep track of the version when in debug?

**Reply by raysan5** [reply]
Personally, I'd start with a 0.1 version until I got something functional.

**Comment by mandar1jn** [comment]
So after a period of kinda not doing anything with this, I decided to give it another shot. I'd like to have the way to go about this nailed down first. One of the first things to think about is how to get started adding Vulkan support. Someone in the discord suggested to split rlgl into a header file called rlgraphics and a .c file with all code. That way I can more easily add vulkan support and ensure that the function definitions stay the same.

**Reply by lvdpower** [reply]
It would be great to implement Vulkan support in order to be able to make games for iOS via MoltenVK, since Apple will soon turn off Opengl. I don't really understand this, since I'm just starting to learn graphics programming, but what I found is a similar graphic game library, converted to Vulkan  https://github.com/danilw/nanovg_vulkan  (I checked it works fine). Take a look, I think you will be able to understand and understand the principle of drawing. Pay attention to the huge fps. This is thanks to Vulkan!

**Reply by raysan5** [reply]
> ...since Apple will soon turn off Opengl.

@lvdpower where did you get that information? Afaik, it's deprecated but there has been no announcement or deadline to stop supporting it. Actually, it will probably break many Apple applications still using OpenGL...

**Reply by truedat101** [reply]
There is no deadline on the end of support for OpenGL.  https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_intro/opengl_intro.html

Depreciated since 10.14.  That means it will still have a compatibility layer, but stuck at GL 3.2 or whatever they left in there.   At some point it is probably removed. The iOS situation is worse probably since GLES deprecated since iOS 12: https://developer.apple.com/documentation/opengles .

**Reply by truedat101** [reply]
On the comment from @lvdpower , the approach proposed sounds good.  Technically GL support isn't turned off, but just not moving forward for any newer GL or GLES features.  If a backend design is made, putting the moltenVK wrapper around it is possible.  

Another approach is MGL : https://github.com/openglonmetal/MGL

This is an experimental project it would appear and has open bugs .  This might be more closely aligned with Raylib if the APIs from OpenGL and extensions can be verified to work.

**Reply by raysan5** [reply]
Adding Vulkan support is not only a considerable project but also would add a considerable maintenance work. I'm afraid I can't assume it.

**Reply by MichaelHoelzl** [reply]
Forget about Vulkan or Metal. WebGPU is maybe a better solution. Sounds like "web", but it was also designed to be used not only for the web but also natively on the operating systems for games or applications.

And here's why...
WebGPU allows platform-independent rendering in the browser and on the desktop, whether Windows, Mac or Linux. It uses the graphics interface in the background, which is available Vulkan, Metal or DirectX. An implementation of WebGPU is available from Mozilla (WGPU) and from Google (Dawn). As this is also the future of 3D rendering in web browsers, this interface will become more and more popular and will be supported by all major operating systems. And it is easier to implement than Vulkan. :-)

**Reply by wwderw** [reply]
I do think the Sokol header(s) have the ability to compile against wgpu-native.  Or atleast that is on the roadmap.  I bring that up just based on what is available in "external".

With SDL as a platform, would we not be able to get to Vulkan via SDL? 

Warning, I know just enough to cause damage, so my ideas could be nonsensical.

**Reply by MichaelHoelzl** [reply]
No SDL would not help with Vulkan. But that's interesting about Sokol and WGPU, I didn't know that before, I'll have to try it out.
But it would make sense if Sokol implemented WebGPU. WebGPU is much simpler than Vulkan and you can use any graphics api, even Apple's Metal. I think WebGPU will replace OpenGL in the future when it comes to cross-platform graphics programming.

**Reply by truedat101** [reply]
Understood, I think it's been said by you at least 6 times on the forums that it isn't in the annual roadmap (every time the roadmap requests are posted this comes up).  As experimental side projects, anyone can try out prototypes with the understanding they would need to maintain it.

**Reply by wwderw** [reply]
@MichaelHoelzl I thought SDL was about to target Vulkan, but it's very likely I'm wrong about it.  I really only use SDL_Render for some things that I do with IMGUI.

Yea, Sokol headers I think are all about the web, so the focus for those implementations are first class.  At least the last time that I had messed with them, but I was also using GL with it.  Rust isn't my thing though, but at least they have the static libs precompiled for wgpu native.

**Comment by m1karii** [comment]
@mandar1jn 
Any updates on this?

**Reply by mandar1jn** [reply]
Yes. I don’t have enough experience and don’t know how I would go about implementing it. Raylib also doesn’t really need it.

**Reply by m1karii** [reply]
It's on raysan if he decides to do it then i guess ...

**Reply by raysan5** [reply]
No plans for now.

**Comment by orcmid** [comment]
Folks who continue to want this should look at how raylib has been abstracted to allow other backends.  This has already been done for SDL.  Look at `raylib/src/platforms`.  The `rcore_template.c` there is also instructive.

**Comment by goofyseeker311** [comment]
vulkan for hw rt. because opengl/webgl does not support hw rt.
not compute ray tracing but actual hardware ray tracing using rt cores or native rt implementation.

**Comment by MayaChen350** [comment]
I really feel like this should be more of a plan than just an idea. I'm not saying to ditch OpenGL entirely, but supporting Vulkan (maybe even by default) would most likely make Raylib more powerful than it has ever been

So I feel like there should be more of a collective effort... Maybe I'm not considering everything though but, by observation, OpenGL is starting to get replaced by Vulkan a lot nowadays... For obvious reasons: OpenGL is getting 'outdated' with Vulkan being around.

(For compatibility issues though, I guess OpenGL should stay optional. Though since Vulkan's API is pretty different from OpenGL's if I recall I don't think only a single or 2 devs can implement it on Raylib without breaking raylib. 

I personally think Raylib needs an actual plannified rewrite)

**Comment by KTRosenberg** [comment]
It would be nice to have a vulkan layer and support moltenvk on mac to get it out of the rut. Raylib is great as potential onboarding software for students, so it would help to support modern APIs (at least simple implementations). Actually, you might want to consider some form of webgpu instead of vulkan for a target to retain web compatibility.

**Reply by raysan5** [reply]
you can check raygpu project: https://github.com/manuel5975p/raygpu

**Reply by KTRosenberg** [reply]
> you can check raygpu project: https://github.com/manuel5975p/raygpu

That looks promising. My concern that it might go out-of-sync. Maybe eventually you’ll have some update and I can use this in the meantime.

**Reply by orcmid** [reply]
So much free advice and, it seems, little understanding of how open-source projects work. 

@mandar1jn made a start, but there seems to be no contribution along the lines by which SDL and other back-ends have been made pluggable.  The existing adaption layers might be instructive.  Take a look at `raylib/src/platforms`.  The file `rcore_template.c` is effectively a checklist of everything that needs to be supplied in creating an alternative underlying platform, such as Vulkan.  The existing cases can also be inspected for additional understanding.

FUNDAMENTALLY: All of the suggestions on this discussion require someone to step up and attempt to do the work.  If there are no takers, expanding the requests and polishing proposals of coverage don't accomplish anything.  The current separation of platforms was accomplished mainly by a single (no-longer around) volunteer.

It takes an itch and competence beyond musings over what's nice to have, along with quarter-backing the development process.  If you want a plan, an onlooker should propose one that they intent to follow :).

**Reply by KTRosenberg** [reply]
> So much free advice and, it seems, little understanding of how open-source projects work.
> 
> @mandar1jn made a start, but there seems to be no contribution along the lines by which SDL and other back-ends have been made pluggable. The existing adaption layers might be instructive. Take a look at `raylib/src/platforms`. The file `rcore_template.c` is effectively a checklist of everything that needs to be supplied in creating an alternative underlying platform, such as Vulkan. The existing cases can also be inspected for additional understanding.
> 
> FUNDAMENTALLY: All of the suggestions on this discussion require someone to step up and attempt to do the work. If there are no takers, expanding the requests and polishing proposals of coverage don't accomplish anything. The current separation of platforms was accomplished mainly by a single (no-longer around) volunteer.
> 
> It takes an itch and competence beyond musings over what's nice to have, along with quarter-backing the development process. If you want a plan, an onlooker should propose one that they intent to follow :).

Sometimes open-source projects are very opinionated on whether contributions versus suggestions are allowed, so at least in my case, I was testing the waters by expressing interest in the feature, so see whether there was demand.

**Reply by orcmid** [reply]
@KTRosenberg 
You might have said so.  Stating your question might have been better.

It might be useful to look at the history of pull requests to see how contributions are appreciated here.

And the bindings are pretty much all volunteer contributions, and so-maintained, pretty much.

**Reply by raysan5** [reply]
@KTRosenberg @orcmid For my part, there is no intention to work on a Vulkan backend for the upcoming future, so I pointed to the alternative version.

**Reply by KTRosenberg** [reply]
> @KTRosenberg @orcmid For my part, there is no intention to work on a Vulkan backend for the upcoming future, so I pointed to the alternative version.

Yes, thank you. I had some thoughts about a metal backend, but if it’s already being done in that other project, maybe it’s best not to spend the time unless there’s interest? (I do have experience with Metal 3, which should be enough in the future if I have bandwidth/if there’s interest. I’d start a new discussion in that case.)

**Reply by raysan5** [reply]
@KTRosenberg Note that the main issue with additions is maintenance, raylib has been maintained for 12.5 years now, if you add some new big piece, how long are you ready to maintain it? how many years? That's the main concern with new additions, so, many times they are discarded, unless I'm willing to take the ownership and responsibility involved.
